### ISSUE TITLE
memcached crash on Debian 9 ( memcached 1.4.33 and similar one on 1.5.0)

### ISSUE THREAD

**Issue by farshidce** [memcached/memcached#299]
we have seen memcached 1.4.33 crash on many of our Debian 9 machines with this segfault
Jul 29 07:53:20 cid1004 kernel: [3952384.853073] memcached[46864]: segfault at 7ee15bfff010 ip 0000560640fb2bdd sp 00007f3f60ea8f30 error 4 in memcached[560640f9e000+23000]
~

Aug 1 10:33:30 cid1002 kernel: [4220877.876325] memcached[45061]: segfault at 7f9433fff010 ip 000055b8bf126bdd sp 00007ff4bf7fdf30 error 4 in memcached[55b8bf112000+23000]

after we upgraded to 1.5.0 (we picked up the Debian package) many of the memcached instances crashed again this time with a new backtrace

memcached[16448]: segfault at 7f016ffff010 ip 000055f569bcbead sp 00007f5faeffcf30 error 4 in memcached[55f569bb5000+26000]

the 1.4.33 crash happened in the course of 2 weeks of uptime but each crash happened in different time. The 1.5.0 crash however happened when few hundred clients disconnected and re-connected tot memcached so its possible the cause of the crash is different too

since these are debian builds I was not able to obtain the backtrace. If someone is more familiar with Debian builds and knows how to extract such info I will be happy to investigate more.

root@cid1010:~# uname -a
Linux cid1010 4.9.0-3-amd64 #1 SMP Debian 4.9.30-1 (2017-06-04) x86_64 GNU/Linux

              total        used        free      shared  buff/cache   available
Mem:           503G         37G        457G        3.7G        8.8G        459G
Swap:          1.9G          0B        1.9G

```
Trying ::1...
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
stats
STAT pid 24938
STAT uptime 1158
STAT time 1503673490
STAT version 1.5.0
STAT libevent 2.0.21-stable
STAT pointer_size 64
STAT rusage_user 60.132000
STAT rusage_system 47.532000
```

memcached is started by systemd
```
[Service]
LimitNOFILE=32768
ExecStart=/usr/bin/numactl --interleave=all /usr/share/memcached/scripts/systemd-memcached-wrapper /srv/etc/%i.conf
```

conf file:
```
-m 360000
-p 11241
-u memcache
-l 0.0.0.0
-c 8192
-d
```

**Comment by dormando** [comment]
where're you getting these builds from? I think 9 has 1.4.33, but where'd 1.5.0 come from?

Any chance you could get an addr2line from the segfault? ie, take the number after ip and run `addr2line -e memcached $num`

Also, just for sanity's sake since the formatting in your free output is tough to read; you are purposefully giving it 360 gigabytes of ram yes?

**Comment by farshidce** [comment]
I picked up the deb package from 
https://packages.debian.org/sid/memcached

I had tried to download the source tarball from the debian.org but addr2line still didn't work. Last I checked was that the memcached binary wasent compiled with flags that addr2line needs , I will keep looking to see if i can get addr2line to work there.

Yes we want to use 360gb and this crash and even the previous one happened when the memcache was already full and was evicting items , not sure if that matters

**Comment by dormando** [comment]
any chance you could run a machine straight from the main source tarball? it's pretty easy to compile and should slot into your existing systemd scripts if you copy over the binary.

**Comment by dormando** [comment]
(but also; you should be able to grab the symbols extras under debian and make addr2line work, so one of those options will hopefully help). I've not heard of any crashes in a while so I hope to resolve this soon.

**Comment by farshidce** [comment]
I will try both but given that this is not easily reproducible so far I will first try to add missing symbols to make add2line work on existing memcached binaries and then use it to map 1.4.33 and 1.5.0 segfaults.

**Comment by dormando** [comment]
good luck. let me know if there's anything I can do in the meantime.

**Comment by dormando** [comment]
also; if I could get copies of full stats output (privately, if that's trouble) (stats, stats items, stats slabs, stats settings) it can be helpful in narrowing down which code paths are problematic. bonus points if they're from an instance that crashed and relatively close to the crash.

thanks!

**Comment by farshidce** [comment]
unfortunately I don't have the stats output from these memcached. we only collect curr_items and throw away the rest.
I am still going through Debian wiki pages to figure out how to add the missing symbols for 1.5.0 but I think i have the right build and symbols for 1.4.33 crash we have seen. I ran info symbols on 5 crashes and they all point to : 
assoc_maintenance_thread + 189 in section .text
(gdb) info symbol 0x14bdd
assoc_maintenance_thread + 189 in section .text
(gdb) info symbol 0x14bdd
assoc_maintenance_thread + 189 in section .text

I ran info symbol on all 1.5.0 crashes as well. I am not still 100% sure about the way I re-built the 1.5.0 symbols. I will update the thread if I find a diff trace.

Reading symbols from memcached...done.
(gdb) info symbol 0x16EAD
do_slabs_stats + 2464 in section .text

**Comment by dormando** [comment]
not sure why you have to rebuild it. The debug symbols are automatically created from the official builds:
https://wiki.debian.org/AutomaticDebugPackages - you have to add the debian-debug repos then install memcached-dbgsym so far as I can tell. Then the addr2line should work?

**Comment by dormando** [comment]
also; you can just grab a stats dump (for all the commands I asked) from any machine that's ideally full and evicting to give me a start.

**Comment by farshidce** [comment]
Yeah it's available for most version but Somehow the dbgsymb version of 1.5.0-1 is missing from all Debian repos (e.g http://debug.mirrors.debian.org/debian-debug/pool/main/m/memcached/)

The dbgsymb version is available for 1.5.0 b1 which is different from 1.5.0-1
I will have to ask the maintainer of dbgsymb package to see if it's available anywhere I can download from

**Comment by dormando** [comment]
ah weird. I saw it in the packages list. what about 1.4.33? you have a segfault of that one you can resolve too

**Comment by farshidce** [comment]
installed memcached-dbgsym  for 1.4.33 and used both gdb and addr2line.

based on sefgault

```
messages.3.gz:Jul 31 15:47:09 cid1006 kernel: [4152954.040931] memcached[47687]: segfault at 7ec58bfff010 ip 000055e1e64d8bdd sp 00007f2742cc1f30 error 4 in memcached[55e1e64c4000+23000]
```

```
root@cid1006:/var/log# dpkg -l | grep memcached
ii  libhashkit-dev                 1.0.18-4.1                     amd64        libmemcached hashing functions and algorithms (development files)
ii  libhashkit2:amd64              1.0.18-4.1                     amd64        libmemcached hashing functions and algorithms
ii  libmemcached-dev               1.0.18-4.1                     amd64        C and C++ client library to the memcached server (development files)
ii  libmemcached11:amd64           1.0.18-4.1                     amd64        C and C++ client library to the memcached server
ii  libmemcachedutil2:amd64        1.0.18-4.1                     amd64        library implementing connection pooling for libmemcached
ii  memcached                      1.4.33-1                       amd64        high-performance memory object caching system
ii  memcached-dbgsym               1.4.33-1                       amd64        Debug symbols for memcached
root@cid1006:/var/log# addr2line -e /usr/bin/memcached 14bdd
./assoc.c:230
root@cid1006:/var/log# addr2line -e /usr/bin/memcached 000055e1e64d8bdd
??:0
root@cid1006:/var/log# gdb /usr/bin/memcached 
GNU gdb (Debian 7.12-6) 7.12.0.20161007-git
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/bin/memcached...Reading symbols from /usr/lib/debug/.build-id/0f/00482ef25da1c88117f24a2326795273b7a7ec.debug...done.
done.
(gdb) info symbol 0x000055e1e64d8bdd
No symbol matches 0x000055e1e64d8bdd.
(gdb) info symbol 0x14bdd
assoc_maintenance_thread + 205 in section .text of /usr/bin/memcached
```

0x14bdd is ip -> 000055e1e64d8bdd minus 55e1e64c4000

**Comment by dormando** [comment]
hrm. that line would be pretty concerning.

any chance I could get that full set of stats from any instance you have in that pool which is presently full and evicting? I can set up a repro (need `stats | stats items | stats slabs | stats settings`)

**Comment by farshidce** [comment]
these stats are from 1.5.0 installation that was converted from 1.4.33 after it crashed few weeks ago
https://gist.github.com/farshidce/8df1529762440eefd7666b0aa6e521f0

**Comment by dormando** [comment]
0 cmd_get? is that instance in use?

and yeah I have a suspicion about memory usage; that's why I was asking for stats from a normal instance that's near full. Are any of the crashed 1.4.33's with 360g refilled or nearly refilled?

**Comment by farshidce** [comment]
this mc ( 1.5.0 ) has been used since it was restarted  due to the crash few days ago

https://gist.github.com/farshidce/2092ab52af08fb018a8f052b94470c90

i also have the "stats" being collected from all 1.5.0 so in case they crash i can submit their logs in case crash occurs again. if it helps i can also send the logs when eviction starts on those 360 GB memcahed instances.

**Comment by dormando** [comment]
You don't have any full 1.4.33's at 360G that you could just telnet to and grab the stats from?

I have a feeling it's this: https://github.com/dormando/memcached/commit/c5b3ecba410477fc2ada57b77aa84d53694ed5a5

branch here:
https://github.com/dormando/memcached/tree/hashoverwhelming

hashmasks are 32bit but some pre-x86-64 code for hash table sizing thinks it's 4 bytes but is actually 8 bytes on a 64bit machine. It would've overflowed and broken either way though. I see in your stats output that almost all items sit in the 96b class.

I don't actually have any machines on hand that can repro this since I need around 300 gigs even if I cut the min slab class size down to 64b.

I'm still auditing the code to see if anything else breaks after 2^32 items are stored, but I'm *pretty* sure everything but the hash table itself was made 64bit safe by now. Some stats output from a full/evicting instance would help confirm that while I devise some tests.

**Comment by farshidce** [comment]
I can downgrade one of the existing 1.5 memcached instances to 1.4.33 and wait few days until eviction starts . This could take few days though 
Based on your theory is it necessary for the instance to have come-get or is it sufficient if there are cotnious set commands on the instance until eviction starts. 

If it helps I can also send the stats from other memcached that use less than 250gb memory and they had not crashed yet .

**Comment by dormando** [comment]
Don't bother downgrading. Sorry, I thought you had some 360's at 1.4.33 still. One of the < 250's would be fine, I can draw a line based on the numbers in those stats.

Sets only should be able to repro, but due to the way hashtables expand you can't just flood it with sets from a benchmark. has to grow a bit more naturally. Though you can add `-o hashpower=32` to the startup options then flood it (or 31?).

**Comment by farshidce** [comment]
ok. also , here are the stats from a memcached with 1.4.33 that uses less RAM and had not crashed

https://gist.github.com/farshidce/7af244044d833694845576d8358b7f2b

**Comment by dormando** [comment]
Thanks! I feel like my patch might do it, if you can spare a machine to try it on.

Also just out of curiosity; do these fill up to some amount before getting gets, or how does your use case work? If you don't mind me asking :)

**Comment by dormando** [comment]
Also for what it's worth; you might notice some mild slowdown, but I don't think your ram capacity is high enough to cause a severe problem. Especially with murmurhash enabled in 1.5.

Getting 64bit hashtables supported is on the TODO list behind a few more big batches of work though. So maybe late this year, or early next, and hopefully before 1TB instances become common :) Under 64bit machines many 64bit hashes are faster than their 32bit equivalents anyway.

**Comment by farshidce** [comment]
yes, we do wait until the cache is at 50% or so of its capacity before the cmd_get traffic starts.

i can apply the patch once i figure out how to re-build it using debian scripts. just to make sure i understand should i apply this patch to the 1.5.0 installations ?

**Comment by dormando** [comment]
Yeah, the 1.5.0. The patch is on top of 1.5.1 but it should apply. If you're fetching the debian source directory it's relatively trivial to either swap source files or add a patch file.

**Comment by farshidce** [comment]
4 memcmached running 1.5.0 crashed and i think the eviction had not begun 

here are the "stats" from them 
https://gist.github.com/farshidce/1543ede69269b5dbc4dfd3d5798c6eed
I do have the stats from them for the past 24 hours and if it helps i can send them privately.

i dont have stats slabs and stats items from them  though.

**Comment by dormando** [comment]
do you have addr2line's? especially of that third one.

**Comment by dormando** [comment]
That third one is confusing; it has zero traffic. What was the time delta between the stats output and crash for that one?

**Comment by farshidce** [comment]
since I don't have the dbgsym build of 1.5.0 I_1  don't know If I can trust the "info symbol"
[6413766.405039] memcached[29289]: segfault at 7f8ae3fff020 ip 000055c239c10ead sp 00007feff8986f30 error 4 in memcached[55c239bfa000+26000]

but this is what I get 

```
Reading symbols from memcached...done.
(gdb) info symbol 0x16EAD
assoc_delete + 13 in section .text
```

I do have a memcached with 1.5.0-1+b4 and if that instance crashes I can get the info symbols on it since 1.5.0.-1b4 dbg build is available

**Comment by farshidce** [comment]
double checking the gist right now ...

**Comment by farshidce** [comment]
I have uploaded stats from three instances that crashed. the stats in the tarball are from the last 12 hours or so .
https://drive.google.com/file/d/0B5OKKYhgSvqsdzE5akk0MmJveVE/view

please ignore the "third" from the ^ gist . it was from an inactive instance.

**Comment by dormando** [comment]
thanks!

So, 3221225472 is the exact cutoff when moving from hashpower 31 to 32. I see all three of those stats logs you gave me die right around there.

The patch I gave you limits it to 32... but not sure why 32 would crash. If you repro a crash with 32 you can just modify the HASHPOWER_MAX value in the patch to 31 and see if that stabilizes.

I'll take a closer look when I can.

**Comment by dormando** [comment]
got a repro of 31b -> 32b hash expansion segfaulting at that line. waiting for a headache to disperse and I'll figure out why.. was pretty sure 32b should work.

**Comment by dormando** [comment]
Believe I fixed it: https://github.com/dormando/memcached/commit/225366101f59e14af0a62b8beb8d1272cb0bf2f6
(pushed to the same branch as before)
thankfully trivial (so far).  If you run this branch it shouldn't crash, at least not in this area.

Also, for what it's worth, it takes a very long time to expand the hash table once above 30b. You should considering monitoring the 'hashpower' stat and adding `-o hashpower=N` to the startup options based on how large it got in recent runs, probably broken down by how much ram each host has.

Partly because it has to keep the old hash table around along with the new one while it's expanding, which gets quite large after 30b. Partly because it just burns CPU for 5-15 minutes going from 31 to 32.

Pretty surprised nobody had gotten more than 3.2b items in memcached before, given this bug is 11 years old! Bound to happen some day.

**Comment by dormando** [comment]
Any luck?

**Comment by farshidce** [comment]
haven't had a chance to apply the patch yet since my Debian skills aren't that great in building from source. I should be able to report back in few days though.

**Comment by farshidce** [comment]
I have deployed the patched version couple of days ago and reverted the memory back to 360 GB.
I should be able to confirm whether crash occurs again or not in a week.

**Comment by dormando** [comment]
good to hear. I need to cut another release soon.

**Comment by dormando** [comment]
how'd it go? :) I should release this this weekend, if you've managed to go over 3.2b objects.

**Comment by farshidce** [comment]
no curr_items is at 2.5 billion now . It should get there in less than a week though. does it help if there are more "gets" or is curr_items the only way to trigger the crash ?

**Comment by dormando** [comment]
curr_items is the only way.

**Comment by farshidce** [comment]
it has surpassed 3.2 now but no eviction yet.  once it starts eviction i will update the thread as well.
```
STAT connection_structures 212
STAT reserved_fds 20
STAT cmd_get 0
STAT cmd_set 14413086019
STAT cmd_flush 0
STAT cmd_touch 0
STAT get_hits 0
STAT get_misses 0
STAT get_expired 0
STAT get_flushed 0
STAT delete_misses 0
STAT delete_hits 0
STAT incr_misses 0
STAT incr_hits 0
STAT decr_misses 0
STAT decr_hits 0
STAT cas_misses 0
STAT cas_hits 0
STAT cas_badval 0
STAT touch_hits 0
STAT touch_misses 0
STAT auth_cmds 0
STAT auth_errors 0
STAT bytes_read 1995118762855
STAT bytes_written 56514914505
STAT limit_maxbytes 377487360000
STAT accepting_conns 1
STAT listen_disabled_num 0
STAT time_in_listen_disabled_us 0
STAT threads 4
STAT conn_yields 50712660
STAT hash_power_level 32
STAT hash_bytes 34359738368
STAT hash_is_expanding 0
STAT slab_reassign_rescues 4031951
STAT slab_reassign_chunk_rescues 0
STAT slab_reassign_evictions_nomem 0
STAT slab_reassign_inline_reclaim 632
STAT slab_reassign_busy_items 55
STAT slab_reassign_busy_deletes 0
STAT slab_reassign_running 0
STAT slabs_moved 1244
STAT lru_crawler_running 1
STAT lru_crawler_starts 92185
STAT lru_maintainer_juggles 4459561986
STAT malloc_fails 0
STAT log_worker_dropped 0
STAT log_worker_written 0
STAT log_watcher_skipped 0
STAT log_watcher_sent 0
STAT bytes 345601420404
STAT curr_items 3645773016
STAT total_items 14417117970
STAT slab_global_page_pool 0
STAT expired_unfetched 0
STAT evicted_unfetched 0
STAT evicted_active 0
STAT evictions 0
STAT reclaimed 0
STAT crawler_reclaimed 0
STAT crawler_items_checked 563440512109
STAT lrutail_reflocked 0
STAT moves_to_cold 6473590922
STAT moves_to_warm 0
STAT moves_within_lru 0
STAT direct_reclaims 0
STAT lru_bumps_dropped 0```

**Comment by dormando** [comment]
`hash_power_level 32` - excellent. looks like it successfully expanded (unless you started it that large?). I'll cut a release this weekend if I can regardless.

Thanks for your patience in getting this figured out! Sorry I left that bug for so long. I'll leave this open until you report successful evictions :P

**Comment by farshidce** [comment]
ongoing evictions for few days and no crash so far :)

STAT auth_errors 0
STAT bytes_read 2342629179875
STAT bytes_written 66338251139
STAT limit_maxbytes 377487360000
STAT accepting_conns 1
STAT listen_disabled_num 0
STAT time_in_listen_disabled_us 0
STAT threads 4
STAT conn_yields 59541931
STAT hash_power_level 32
STAT hash_bytes 34359738368
STAT hash_is_expanding 0
STAT slab_reassign_rescues 5025597
STAT slab_reassign_chunk_rescues 0
STAT slab_reassign_evictions_nomem 615597
STAT slab_reassign_inline_reclaim 443100
STAT slab_reassign_busy_items 615687
STAT slab_reassign_busy_deletes 0
STAT slab_reassign_running 0
STAT slabs_moved 1604
STAT lru_crawler_running 1
STAT lru_crawler_starts 106429
STAT lru_maintainer_juggles 5211087955
STAT malloc_fails 0
STAT log_worker_dropped 0
STAT log_worker_written 0
STAT log_watcher_skipped 0
STAT log_watcher_sent 0
STAT bytes 349840389083
STAT curr_items 3693186283
STAT total_items 16925237992
STAT slab_global_page_pool 0
STAT expired_unfetched 0
STAT evicted_unfetched 408186948
STAT evicted_active 0
STAT evictions 408186948
STAT reclaimed 0
STAT crawler_reclaimed 0
STAT crawler_items_checked 713096544837
STAT lrutail_reflocked 0
STAT moves_to_cold 7594040314
STAT moves_to_warm 0
STAT moves_within_lru 0
STAT direct_reclaims 408186948
STAT lru_bumps_dropped 0
END

**Comment by dormando** [comment]
that is a lot of items. thanks for verifying! I didn't get a release over the weekend but one will happen soon. I should stack a few more small fixes in first anyway.

**Comment by dormando** [comment]
released in 1.5.2. thanks again
